cpp/src/tsne/tsne_runner.cuh

```diff
@@ -0,0 +1,202 @@
+      exaggeration_iter(exaggeration_iter_),
+      min_gain(min_gain_),
+      pre_learning_rate(pre_learning_rate_),
+      post_learning_rate(post_learning_rate_),
+      max_iter(max_iter_),
+      min_grad_norm(min_grad_norm_),
+      pre_momentum(pre_momentum_),
+      post_momentum(post_momentum_),
+      random_state(random_state_),
+      verbosity(verbosity_),
+      initialize_embeddings(initialize_embeddings_),
+      barnes_hut(barnes_hut_),
+      COO_Matrix(handle_.get_device_allocator(), handle_.get_stream()) {
+    this->n = input.n;
+    this->p = input.d;
+    this->Y = input.y;
+
+    ML::Logger::get().setLevel(verbosity);
+    if (dim > 2 and barnes_hut) {
+      barnes_hut = false;
+      CUML_LOG_WARN(
+        "Barnes Hut only works for dim == 2. Switching to exact solution.");
+    }
+    if (n_neighbors > n) n_neighbors = n;
+    if (n_neighbors > 1023) {
+      CUML_LOG_WARN("FAISS only supports maximum n_neighbors = 1023.");
+      n_neighbors = 1023;
+    }
+    // Perplexity must be less than number of datapoints
+    // "How to Use t-SNE Effectively" https://distill.pub/2016/misread-tsne/
+    if (perplexity > n) perplexity = n;
+
+    CUML_LOG_DEBUG("Data size = (%d, %d) with dim = %d perplexity = %f", n, p,
+                   dim, perplexity);
+    if (perplexity < 5 or perplexity > 50)
+      CUML_LOG_WARN(
+        "Perplexity should be within ranges (5, 50). Your results might be a"
+        " bit strange...");
+    if (n_neighbors < perplexity * 3.0f)
+      CUML_LOG_WARN(
+        "# of Nearest Neighbors should be at least 3 * perplexity. Your results"
+        " might be a bit strange...");
+  }
+
+  void run() {
+    distance_and_perplexity();
+
+    const int NNZ = COO_Matrix.nnz;
+    float *VAL = COO_Matrix.vals();
+    const int *COL = COO_Matrix.cols();
+    const int *ROW = COO_Matrix.rows();
+    //---------------------------------------------------
+
+    if (barnes_hut) {
+      TSNE::Barnes_Hut(VAL, COL, ROW, NNZ, handle, Y, n, theta, epssq,
+                       early_exaggeration, exaggeration_iter, min_gain,
+                       pre_learning_rate, post_learning_rate, max_iter,
+                       min_grad_norm, pre_momentum, post_momentum, random_state,
+                       initialize_embeddings);
+    } else {
+      TSNE::Exact_TSNE(VAL, COL, ROW, NNZ, handle, Y, n, dim,
+                       early_exaggeration, exaggeration_iter, min_gain,
+                       pre_learning_rate, post_learning_rate, max_iter,
+                       min_grad_norm, pre_momentum, post_momentum, random_state,
+                       initialize_embeddings);
+    }
+  }
+
+ private:
+  void distance_and_perplexity() {
+    START_TIMER;
+
+    //---------------------------------------------------
+    // Get distances
+    CUML_LOG_DEBUG("Getting distances.");
+
+    auto stream = handle.get_stream();
+
+    rmm::device_uvector<knn_value_idx> indices(0, stream);
```

Comment:
Happy to see an example of this in cuml. Very happy to see `device_buffer` get removed.